ApplicationLibCode/ModelVisualization/Intersections/RivPolylineIntersectionGeometryGenerator.cpp

```diff
@@ -0,0 +1,312 @@
+    // can be calculated using Pythagoras theorem. Where a segment is defined between p1 and p2,
+    // which implies p1 is origin of the local coordinate system uz.
+    //
+    // For a segment a UTM coordinate (x_utm,y_utm,z_utm) converts to u = sqrt(x^2 + y^2) and z = z.
+    // Where x, y and z are local vertex coordinates, after subtracting the (x_utm, y_utm, z_utm)-values
+    // for p1 from the vertex UTM-coordinates.
+    while ( pointIdx < numPoints - 1 )
+    {
+        size_t nextPointIdx = RivSectionFlattener::indexToNextValidPoint( m_polylineUtm, zAxisDirection, pointIdx );
+        if ( nextPointIdx == size_t( -1 ) || nextPointIdx >= m_polylineUtm.size() )
+        {
+            break;
+        }
+
+        // Start and end point of polyline segment in UTM-coordinates
+        const cvf::Vec3d p1 = m_polylineUtm[pointIdx];
+        const cvf::Vec3d p2 = m_polylineUtm[nextPointIdx];
+
+        // Get cell candidates for the polyline segment (subset of global cells)
+        std::vector<size_t> columnCellCandidates =
+            createPolylineSegmentCellCandidates( *m_hexGrid, p1, p2, maxHeightVec, topDepth, bottomDepth );
+
+        // Plane for the polyline segment
+        cvf::Plane plane;
+        plane.setFromPoints( p1, p2, p2 + maxHeightVec );
+
+        // Planes parallel to z-axis for p1 and p2, to prevent triangles outside the polyline segment
+        cvf::Plane p1Plane;
+        p1Plane.setFromPoints( p1, p1 + maxHeightVec, p1 + plane.normal() );
+        cvf::Plane p2Plane;
+        p2Plane.setFromPoints( p2, p2 + maxHeightVec, p2 - plane.normal() );
+
+        // Placeholder for triangle vertices per cell
+        std::vector<caf::HexGridIntersectionTools::ClipVx> hexPlaneCutTriangleVxes;
+        hexPlaneCutTriangleVxes.reserve( 5 * 3 );
+        std::vector<int> cellFaceForEachTriangleEdge;
+        cellFaceForEachTriangleEdge.reserve( 5 * 3 );
+        cvf::Vec3d cellCorners[8];
+        size_t     cornerIndices[8];
+
+        // Polyline segment data
+        std::vector<cvf::uint> polygonIndices        = {};
+        std::vector<cvf::uint> verticesPerPolygon    = {};
+        std::vector<cvf::uint> polygonToCellIndexMap = {};
+
+        // Welder for segment vertices
+        // - Number of buckets is size of columnCellCandidates to get a dynamic number of buckets usable for respective segment.
+        cvf::VertexWelder segmentVertexWelder;
+        const cvf::uint   numWelderBuckets = std::max( static_cast<cvf::uint>( columnCellCandidates.size() ), cvf::uint( 1 ) );
+        segmentVertexWelder.initialize( weldingDistance, weldingCellSize, numWelderBuckets );
+
+        // Intersection per grid cell - transform from set of triangles to polygon for cell
+        for ( const auto globalCellIdx : columnCellCandidates )
+        {
+            if ( ( visibleCells != nullptr ) && ( ( *visibleCells )[globalCellIdx] == 0 ) ) continue;
```

Comment:
Maybe this is easier to read
if (  visibleCells  && (( *visibleCells )[globalCellIdx] == 0 ) ) continue;